src/vm/amd64/umthunkstub.S

```diff
@@ -0,0 +1,373 @@
+        mov             METHODDESC_REGISTER, [rbp + UMThunkStubAMD64_METHODDESC_SPILL_OFFSET]
+
+        jmp             LOCAL_LABEL(InCooperativeMode)
+
+LOCAL_LABEL(UMThunkStub_CopyStackArgs):
+        // rax = cbStackArgs (with 20h for register args subtracted out already)
+
+        sub             rsp, rax
+        and             rsp, -16
+
+        // rax = number of bytes
+
+        lea             rdi, [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET + SIZEOF_MAX_OUTGOING_ARGUMENT_HOMES] 
+        lea             rsi, [rsp + UMThunkStubAMD64_CALLEE_SCRATCH_SIZE]
+
+LOCAL_LABEL(CopyLoop):
+        // rax = number of bytes
+        // rdi = src
+        // rsi = dest
+        // rdx = sratch
+
+        add             rax, -8
+        mov             rdx, [rdi + rax]
+        mov             [rsi + rax], rdx
+        jnz             LOCAL_LABEL(CopyLoop)
+
+        mov             rdi, [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET +  0h]
+        mov             rsi, [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET +  8h]
+        mov             rdx, [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET + 10h]
+        
+        jmp             LOCAL_LABEL(UMThunkStub_ArgumentsSetup)
+
+LOCAL_LABEL(WrongAppDomain):
+        //
+        // call our helper to perform the AD transtion 
+        //
+        mov             rdi, METHODDESC_REGISTER
+        lea             rdx,  [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET]
+        mov             rax, [METHODDESC_REGISTER + OFFSETOF__UMEntryThunk__m_pUMThunkMarshInfo]
+        mov             rcx, [rax + OFFSETOF__UMThunkMarshInfo__m_cbActualArgSize]
+        mov             rsi,  [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET +  8h] 
+        call            C_FUNC(UM2MDoADCallBack)
+
+        // restore return value
+        mov             rax,  [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET +  0h]
+        movdqa          xmm0, [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET +  0h]
+
+        jmp             LOCAL_LABEL(PostCall)
+
+NESTED_END UMThunkStub, _TEXT
+
+//
+// EXTERN_C void __stdcall UM2MThunk_WrapperHelper(
+//       void *pThunkArgs,               // rdi
+//       int argLen,                     // rsi
+//       void *pAddr,                    // rdx            // not used
+//       UMEntryThunk *pEntryThunk,      // rcx
+//       Thread *pThread);               // r8
+//
+NESTED_ENTRY UM2MThunk_WrapperHelper, _TEXT, NoHandler
+
+// number of integer registers saved in prologue
+#define UM2MThunk_WrapperHelper_NUM_REG_PUSHES 3
+
+#define UM2MThunk_WrapperHelper_CALLEE_SCRATCH_SIZE SIZEOF_MAX_OUTGOING_ARGUMENT_HOMES
+#define UM2MThunk_WrapperHelper_STACK_FRAME_SIZE 0x48
+
+#define UM2MThunk_WrapperHelper_FRAME_OFFSET UM2MThunk_WrapperHelper_CALLEE_SCRATCH_SIZE
+#define UM2MThunk_WrapperHelper_FIXED_STACK_ALLOC_SIZE (UM2MThunk_WrapperHelper_STACK_FRAME_SIZE - (UM2MThunk_WrapperHelper_NUM_REG_PUSHES * 8))
+
+        push_nonvol_reg r9
```

Comment:
I agree, it shouldn't be needed.  Done and adjusted the defines.  Also, there was the mov METHOEDDESC_REGISTER, r9 below, which should have been rcx.